letsencrypt/client/auth_handler.py

```diff
@@ -26,186 +24,212 @@ class AuthHandler(object):  # pylint: disable=too-many-instance-attributes
             logging.info("Attempting to clean up outstanding challenges...")
-            for dom in self.domains:
-                self._cleanup_challenges(dom)
+            self._cleanup_challenges()
+            raise
+
+        assert len(cont_resp) == len(self.cont_c)
+        assert len(dv_resp) == len(self.dv_c)
 
-            raise errors.LetsEncryptAuthHandlerError(
-                "Unable to perform challenges")
+        return cont_resp, dv_resp
 
-        logging.info("Ready for verification...")
+    def _respond(self, cont_resp, dv_resp, best_effort):
+        """Send/Receive confirmation of all challenges.
 
-        # Assemble Responses
-        if cont_resp:
-            self._assign_responses(cont_resp, self.cont_c)
-        if dv_resp:
-            self._assign_responses(dv_resp, self.dv_c)
+        .. note:: This method also cleans up the auth_handler state.
 
-    def _assign_responses(self, flat_list, ichall_dict):
-        """Assign responses from flat_list back to the Indexed dicts.
+        """
+        # TODO: chall_update is a dirty hack to get around acme-spec #105
+        chall_update = dict()
+        active_achalls = []
+        active_achalls.extend(
+            self._send_responses(self.dv_c, dv_resp, chall_update))
+        active_achalls.extend(
+            self._send_responses(self.cont_c, cont_resp, chall_update))
+
+        # Check for updated status...
+        self._poll_challenges(chall_update, best_effort)
+        # This removes challenges from self.dv_c and self.cont_c
+        self._cleanup_challenges(active_achalls)
+
+    def _send_responses(self, achalls, resps, chall_update):
+        """Send responses and make sure errors are handled.
 
-        :param list flat_list: flat_list of responses from an IAuthenticator
-        :param dict ichall_dict: Master dict mapping all domains to a list of
-            their associated 'continuity' and 'dv' Indexed challenges, or their
-            :class:`letsencrypt.client.achallenges.Indexed` list
+        :param dict chall_update: parameter that is updated to hold
+            authzr -> list of outstanding solved annotated challenges
 
         """
-        flat_index = 0
-        for dom in self.domains:
-            for ichall in ichall_dict[dom]:
-                self.responses[dom][ichall.index] = flat_list[flat_index]
-                flat_index += 1
+        active_achalls = []
+        for achall, resp in itertools.izip(achalls, resps):
+            # Don't send challenges for None and False authenticator responses
+            if resp:
+                self.network.answer_challenge(achall.challb, resp)
+                active_achalls.append(achall)
+                if achall.domain in chall_update:
+                    chall_update[achall.domain].append(achall)
+                else:
+                    chall_update[achall.domain] = [achall]
+
+        return active_achalls
+
+    def _poll_challenges(
```

Comment:
My thoughts are addressed in the acme-spec comment.  I would really like to apply logic to failure conditions.